internal/rbd/rbd_util.go

```diff
@@ -197,17 +200,106 @@ type migrationVolID struct {
 	clusterID string
 }
 
-var supportedFeatures = map[string]imageFeature{
-	librbd.FeatureNameLayering: {
-		needRbdNbd: false,
-	},
-	librbd.FeatureNameExclusiveLock: {
-		needRbdNbd: true,
-	},
-	librbd.FeatureNameJournaling: {
-		needRbdNbd: true,
-		dependsOn:  []string{librbd.FeatureNameExclusiveLock},
-	},
+var (
+	supportedFeatures = map[string]imageFeature{
+		librbd.FeatureNameLayering: {
+			needRbdNbd: false,
+		},
+		librbd.FeatureNameExclusiveLock: {
+			needRbdNbd: true,
+		},
+		librbd.FeatureNameJournaling: {
+			needRbdNbd: true,
+			dependsOn:  []string{librbd.FeatureNameExclusiveLock},
+		},
+	}
+
+	krbdSupportedFeaturesAttr string
+)
+
+// getKrbdSupportedFeatures load the module if needed and return supported
+// features attribute as a string.
+func getKrbdSupportedFeatures() (string, error) {
+	// Added to avoid issues for non-root users who will not have access to
+	// modprobe the module, example CI running unit tests.
+	if krbdSupportedFeaturesAttr != "" {
+		return krbdSupportedFeaturesAttr, nil
+	}
+
+	// check if the module is loaded or compiled in
+	_, err := os.Stat(krbdSupportedFeaturesFile)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			log.ErrorLogMsg("stat on: %s failed, err: %v", krbdSupportedFeaturesFile, err)
+
+			return "", err
+		}
+		// try to load the module
+		_, _, err = util.ExecCommand(context.TODO(), "modprobe", rbdDefaultMounter)
+		if err != nil {
+			log.WarningLogMsg("modprobe failed with error %v", err)
+
+			return "", err
+		}
+	}
+	val, err := ioutil.ReadFile(krbdSupportedFeaturesFile)
+	if err != nil {
+		log.ErrorLogMsg("ReadFile %s failed, err: %v", krbdSupportedFeaturesFile, err)
+
+		return "", err
+	}
+
+	return strings.TrimSuffix(string(val), "\n"), nil
+}
+
+// hexStringToInteger convert hex value to uint.
+func hexStringToInteger(hexString string) (uint64, error) {
+	// trim 0x prefix
+	numberStr := strings.TrimPrefix(strings.ToLower(hexString), "0x")
+
+	output, err := strconv.ParseUint(numberStr, 16, 64)
+	if err != nil {
+		log.ErrorLogMsg("strconv.ParseUint(%s) failed, err: %v", numberStr, err)
+
+		return 0, err
+	}
+
+	return output, nil
+}
+
+// isKrbdFeatureSupported checks if a given Image Feature is supported by krbd
+// driver or not.
+func isKrbdFeatureSupported(fname string, imageFeatureSet librbd.FeatureSet) (bool, error) {
+	supported := false
+	attr, err := getKrbdSupportedFeatures()
+	if err != nil {
+		return supported, err
+	}
+
+	features, err := hexStringToInteger(attr)
+	if err != nil {
+		return supported, err
+	}
+
+	found := false
+	for _, featureName := range imageFeatureSet.Names() {
+		if !strings.EqualFold(fname, featureName) {
+			continue
+		}
+		found = true
+		if (uint64(imageFeatureSet) & features) != 0 {
```

Comment:
this does not need to be inside the for-loop?

Maybe only list the features that are not supported in case there are any? Do not run the for-loop when all required features are supported.

You might be able to get the missing features with something like this:
```
librbd.FeatureSet(uint64(imageFeatureSet) & ~features).Names()
```